New Redis Client API

cescoffier opened the pull request:
The current Redis extension does not provide all the possibilities Redis offers, and the API is not easy to use.

This PR provides a new high-level and type-safe Redis API that covers the previous API's shortcomings.

The previous API is still available but deprecated.

**Breaking Changes**

Unfortunately, while refactoring the extension, I had to break a few things:

- Host Providers are now identified using `@Identifier` instead of `@Named` (following @mkouba's advice)
- It is not possible to create dynamic clients - this was done to support pub/sub, which is now supported by the API

**What's missing (and will come later)**

The PR is already big enough, and some parts will be provided later without breakage:

- the commands from the `stream` group are not yet provided (you can use custom commands for the time being)
- scan operations should provide a flattened stream (so instead of retrieving batches, retrieves the element individually) 
- metrics have been left off for now. It's possible to implement a micrometer binder (work in progress)
- the `LCS` command does not provide the `IDX` option


**Related issues**

- Fix https://github.com/quarkusio/quarkus/issues/24056
- Fix https://github.com/quarkusio/quarkus/issues/21108
-

cescoffier committed:
New Redis API

machi1990 commented:
Thanks @cescoffier for this. Very nice improvements.

I'll have a proper look in the next couple of days.

ebullient commented:
Is it too much to ask to bring the appropriate doc templates into your new docs? 
https://github.com/quarkusio/quarkus/pull/25354

docs/src/main/asciidoc/redis.adoc -- Is that meant to be a tutorial? (I believe so.. )
docs/src/main/asciidoc/redis-reference.adoc -- this feels like a combination of a how-to (installation), and a reference (how the extension works and config options, etc.) .. can we split them so we get the docs roughly better in the first pass? (I am happy to help in that regard.. )

machi1990 commented on docs/src/main/asciidoc/redis-reference.adoc:
```diff
@@ -3,99 +3,646 @@ This guide is maintained in the main Quarkus repository
+
+To help you select the suitable API for you, here are some recommendations:
+
+* If you are building an imperative (_classic_) Quarkus application integrating with Redis: use `io.quarkiverse.redis.datasource.api.RedisDataSource`.
```
```suggestion
* If you are building an imperative (_classic_) Quarkus application integrating with Redis: use `io.quarkus.redis.datasource.api.RedisDataSource`.
```

The package name change to `io.quarkus.redis.datasource.api` has to be applied throughout 
```suggestion
* If you are building an imperative (_classic_) Quarkus application integrating with Redis: use `io.quarkiverse.redis.datasource.api.RedisDataSource`.
```

cescoffier replied:
Great catch! I initially planned to contribute the new client to quarkiverse.

machi1990 commented on docs/src/main/asciidoc/redis-reference.adoc:
```diff
@@ -3,99 +3,646 @@ This guide is maintained in the main Quarkus repository
+
+=== Cluster Mode
+
+When using Redis in cluster mode, you need to pass multiple _host urls_, configure the client type to `CLUSTER` and configure the `replicas` mode:
```
```suggestion
When using Redis in cluster mode, you need to pass multiple _host urls_, configure the client type to `cluster` and configure the `replicas` mode:
```

machi1990 reviewed (approved):
I left some minor doc suggestions, otherwise lgtm. 

Will it be possible to add a migration note as well as there are some breaking changes?

I love the direction that this is taking with typesafety, repository style api, thanks @cescoffier for the PR.

cescoffier committed:
Update to Vert.x 4.3.1

Vert.x 4.3.1 Redis clients has a few API breakage. This commit works around them.

In addition, Redis 7.0 dropped the host and post commands.
These commands are not exposed in the new API and are deprecated in the old one.
